### PR TITLE
[Fix #16919] Add bash completion for `docker network`

### LINKED ISSUE #16919: Bash completion for `docker network`
This is to document that `docker network` is missing in bash completion.
I will add it.

### PR DESCRIPTION
**Pull request by albers**
This adds bash completion for `docker network` and subcommands.
ref. #16919

As this is the second command with real subcommands (`docker volume` being the first), I had to do a refactoring first.

ping @jfrazelle @tianon for review
I'm not sure about the completion for `--driver`. Are the values OK like this?

Question: What is the Go format expression that evaulates to true if a container has a given `.NetworkSettings.Networks` value? Something like `.NetworkSettings.Networks contains "bridge"`

### CODE REVIEW

**Comment by vdemeester** [comment]
Tried it out real quick. There is only one strange behaviour with `--help` but I don't know if it's already like that or not (and maybe it's because of [that](https://github.com/docker/docker/pull/16922/files#diff-754420541b12dbcbbea00825942a4635R226)).

```bash
$ docker network --help                                                                                                                  [52/1028]
connect     create      disconnect  inspect     ls          rm          
$ docker network --help ls
docker: 'network' is not a docker command.
See 'docker --help'.
$ docker --help 
attach   cp       diff     export   images   inspect  login    network  ps       rename   rmi      search   stop     unpause  wait     
build    create   events   help     import   kill     logout   pause    pull     restart  run      start    tag      version  
commit   daemon   exec     history  info     load     logs     port     push     rm       save     stats    top      volume   
$ docker --help network 
connect     create      disconnect  inspect     ls          rm          
$ docker --help network ls 
Usage: docker [OPTIONS] COMMAND [arg...]
       docker daemon [ --help | ... ]
       docker [ --help | -v | --version ]

A self-sufficient runtime for containers.

Options:

  --config=~/.docker              Location of client config files

```

Apart from that, LGTM 🐸

File: contrib/completion/bash/docker [hunk 1]
```diff
@@ -138,6 +138,10 @@ __docker_containers_and_images() {
 	COMPREPLY+=( "${containers[@]}" )
 }
 
+__docker_networks() {
+	COMPREPLY=( $(compgen -W "$(__docker_q network ls | awk 'NR>1 {print $2}')" -- "$cur") )
```
Comment (mrjana) [inline contrib/completion/bash/docker]: Can we also try to complete network IDs in addition to names? It's useful and is consistent with what we do for containers (we complete on both IDs and names).

File: contrib/completion/bash/docker [hunk 2]
```diff
@@ -965,6 +998,98 @@ _docker_logs() {
...
+	case "$cur" in
+		-*)
+			COMPREPLY=( $( compgen -W "--driver -d --help" -- "$cur" ) )
+			;;
+	esac
+}
+
+_docker_network_disconnect() {
+	# TODO disconnect should only complete running containers connected
+	# to the specified network.
```
Comment (mrjana) [inline contrib/completion/bash/docker]: We only complete running containers for `docker stop`. Can we try reusing that to only complete running containers here or were you planning to do that later?

**Comment by mrjana** [comment]
@albers Overall the completion changes LGTM. Just two minor comments. Feel free to add them in a subsequent PR

**Comment by tiborvass** [comment]
I wonder if we should wait for #16910 so we can update the flags again?

**Comment by mrjana** [comment]
@tiborvass Yes I think it is a good idea.

Also @albers the completion list for drivers is fine for now. The only additional thing that we can do is to list all known plugins but currently there is no easy way to get it because you need to inspect all previous networks and find what the driver is and include it in the completion list.

File: contrib/completion/bash/docker (hunk 1 above)
Comment (albers) [inline contrib/completion/bash/docker]: Yes, sure. I was in doubt whether IDs are really useful and opted against them. How about making their inclusion configurable with the env var used for containers?

File: contrib/completion/bash/docker (hunk 2 above)
Comment (albers) [inline contrib/completion/bash/docker]: Sorry, I can't follow you here. Documentation states that both `connect` and `disconnect` work on _running_ containers, so I only complete those here.
The comment is about a further restriction that could be applied in `disconnect`. For thar I need help, see my question in OP.

**Comment by albers** [comment]
I like the idea of extracting the drivers from the existing networks. This might become too slow, though as it requires iterating over and inspecting all networks.
I have no idea how many networks I should expect on a host. Is the querying approach feasable? WDYT?

**Comment by mavenugo** [comment]
@albers even if we go with the approach of extracting drivers from existing networks, sometimes some of the drivers will not even show up if the user has not used it atleast once. That kind of defeats the purpose of command completion. It wont show up the first time and it might show up  later after the first use.

**Comment by albers** [comment]
ah, yes, of course.
So this approach is no big deal.
Do network drivers show up in`docker info`?

**Comment by mavenugo** [comment]
@albers i think this is a general problem with plugins in general. they wont show up until it is used for atleast 1 time.

File: contrib/completion/bash/docker (hunk 1 above)
Comment (mrjana) [inline contrib/completion/bash/docker]: Sure that would be a good idea

File: contrib/completion/bash/docker (hunk 2 above)
Comment (mrjana) [inline contrib/completion/bash/docker]: Ah, I didn't realize you were already handling the case of running containers in `connect` case. If you want to see complete only the containers which use a particular network you can use the following
```
/docker inspect --format '{{range $i, $net := .NetworkSettings.Networks}} {{if (eq $net "bridge")}} true {{else}} false {{end}} {{end}}' 9ee4fd9a1224611
```
I tested this. It works :-)

**Comment by thaJeztah** [comment]
@albers @mavenugo see https://github.com/docker/docker/issues/16205 for a proposal to add the available volume drivers to docker info.

@mavenugo Perhaps you add your thoughts there w.r.t. network drivers?

**Comment by jfrazelle** [comment]
LGTM

**Comment by jfrazelle** [comment]
I think the current approach is fine for now

**Comment by mrjana** [comment]
LGTM from me again.

@albers If you want to improve the container completion during disconnect in the future here's the example inspect template code that should work
```
docker inspect --format '{{range $i, $net := .NetworkSettings.Networks}} {{if (eq $net "bridge")}} true {{else}} false {{end}} {{end}}' 9ee4fd9a1224611
```

**Comment by albers** [comment]
@vdemeester Thanks for evaluating.
Sorry, I could not respond earlier.
Yes, `--help` is a strange option because it applies to all levels.

Completion could be updated not to complete commands or subcommands if `--help` were given at a previous level. WDYT?

**Comment by vdemeester** [comment]
@albers hum that would be cool, but it might be overkill (in terms of complexity to implement) compared to the number of case user will hit that though :stuck_out_tongue:.

**Comment by albers** [comment]
@mrjana Thanks for reviewing and for your comments.

The go template is not perfect for containers that are connected to more than one network. It will produce either `true` or `false` for _each_ of the networks.
It might work for me if I remove the `else` branch and only check the output for being `true`, though.

Is there a way to say *.NetworkSettings.Networks contains a value of "foo"* in go template syntax?

File: contrib/completion/bash/docker (hunk 1 above)
Comment (albers) [inline contrib/completion/bash/docker]: @mrjana See #17270

**Pull request merged**